File: Chapter11/com/packtpub/spark/Exercise11_03/Exercise11_03.scala
package com.packtpub.spark.Exercise11_03

import org.apache.spark.sql.types.{StringType, StructField, StructType}
import org.apache.spark.sql.{Row, SparkSession}

object Exercise11_03 {

  def main(args: Array[String]): Unit = {

    // Build a SparkSession in Local Mode
    val spark = SparkSession
      .builder()
      .master("local[2]")
      .appName("My Spark App")
      .getOrCreate()

    val categorized_animals = Seq(Row("dog", "pet"),
                                  Row("cat", "pet"),
                                  Row("bear", "wild"),
                                  Row("cat", "pet"),
                                  Row("cat", "pet"))

    val schema = List(
      StructField("name", StringType, nullable = true),
      StructField("category", StringType, nullable = true)
    )

    val animalDataRDD = spark.sparkContext.parallelize(categorized_animals)

    val animalData = spark.createDataFrame(animalDataRDD, StructType(schema))

    // filter out any records where name column starts with "c"
    val nonCats = animalData.filter("name not like 'c%'")

    // or you can use the where method, which is an alias on the filter method
    val nonCatsTwo = animalData.where("name != 'cat'")
    nonCatsTwo.show()

    nonCats.foreach(animal => {
      println(s"I am a ${animal.get(0)}, which does not start with a 'c'.")
    })

    val nonPets = animalData.filter("category != 'pet'")

    nonPets.foreach(animal => {
      println(s"I am no pet, for I am a ${animal.get(1)} ${animal.get(0)}!")
    })

  }

}

File: Chapter11/com/packtpub/spark/Exercise11_03.py
from pyspark.sql import SparkSession
from pyspark.sql import Row
from pyspark.sql import SQLContext


# Function to read data in RDD
def read(animal): print("I am no pet, for I am a " + animal['category'], animal['name'] + "!")


# Function to showcase filtered data
def read_non_cats(animal): print("I am a " + animal['name'] + ", which does not start with a 'c'.")


# Create a Spark Session
spark = SparkSession\
    .builder\
    .appName("My Spark App")\
    .master("local[2]")\
    .getOrCreate()

sc = spark.sparkContext
sqlContext = SQLContext(sc)

# create a sample set of data as an RDD
categorized_animals = [("dog", "pet"), ("cat", "pet"), ("bear", "wild"), ("cat", "pet"), ("cat", "pet")]
animalDataRDD = sc.parallelize(categorized_animals)
animals = animalDataRDD.map(lambda x: Row(name=x[0], category=x[1]))
animalsDF = sqlContext.createDataFrame(animals)

# Filter out animal names that start with 'c'
nonCats = animalsDF.filter("name not like 'c%'")
nonCats.foreach(read_non_cats)

# Filter out animal names that are not 'cat'
nonCatsTwo = animalsDF.where("name != 'cat'")

# Filter out animals that are not in the 'pet' category
nonPets = animalsDF.filter("category != 'pet'")
nonPets.foreach(read)
